kbase-extension/static/kbase/js/common/cellComponents/filePathWidget.js

```diff
@@ -23,21 +23,101 @@ define([
             model = Props.make(),
+            /**
+             * Internal data model
+             * The data model / view model is a structure that maintains the overall internal
+             * state of this widget. It holds:
+             *  - all input widgets in the cell
+             *  - all values, kept in an order that's useful for the cell
+             *  - all row information, in a way that's (generally) easily accessible and easy to
+             *    update.
+             *  - each row gets its own immutible unique id, which is used for widget communication and
+             *    management
+             *  - row order, since we can delete from anywhere, can shift and wobble, so that's not helpful.
+             *    instead, we map from rowId -> row order, and update that on insertions and deletions.
+             * In general, we need to support 3 actions.
+             * 1. add row
+             *  - rows get added to the end, new id, new object for data. O(1)
+             * 2. update data
+             *  - when a field widget updates its data, we get a message with row id, parameter id, and new
+             *    value. Mapping from row id -> row order makes this O(1)
+             * 3. push data to main widget (configure tab)
+             *  - we maintain data in the right order with the right set of values, so this is O(1)
+             * 4. delete row
+             *  - rows can get deleted from anywhere, which breaks all the ordering. We need to run a loop
+             *    to figure out the new order after the one that was deleted, which makes this O(n)
+             *
+             * dataModel structure:
+             * This is a simple object with a few keys:
+             * rowOrder: array of unique ids, this is the row ids in the order they should appear on
+             *   screen, thus the order the data should be stored in
+             * rowIdToIndex: mapping from unique rowId to the index in rowOrder. Mainly used on row
+             *   updates
+             * dataValues: an array of key-value-pairs, parameter id -> parameter value
+             * widgets: mapping from rowId to the array of widget objects in that row, so we can stop
+             *   and remove them all.
+             *
+             * example:
+             * {
+             *   rowOrder: ['rowA', 'rowB'],
+             *   rowIdToIndex: {
+             *      rowA: 0,
+             *      rowB: 1
+             *   },
+             *   dataValues: [{
+             *      param1: 'value1',
+             *      param2: 'value2
+             *   }, {
+             *      param1: 'value3',
+             *      param2: 'value4'
+             *   }],
+             *   widgets: {
+             *      rowA: [Object[], Object[]],
+             *      rowB: [Object[], Object[]]
+             *   }
+             * }
+             *
+             * Why do something this complex? We might have 100+ rows, and sifting through to update
```

Comment:
I find this a far more logical way of doing things that faffing around with searching arrays. Why not make life easy with an index?!

Reply:
Following on from that thought, I think it would be easier to keep all data connected to a certain row ID stored in an object under that ID, and then maintain an array with the row IDs in order, i.e.

```
dataModel.byRowId = {
    'rowA': {
        params: { param1: 'val1', ... },
        widgets: [Object[], Object[]]
    },
    'rowB': {
        params: { param1: 'val2', ... },
        widgets: [Object[], Object[]]
    }
}

dataModel.rowOrder = ['rowA', 'rowB']

// generate dataValues:
const dataValues = dataModel.rowOrder
  .filter((elem) => { return elem })
  .map((elem) => { return dataModel.ix[elem].params })

// add a new item
dataModel.rowOrder.push('rowC')
dataModel.byRowId['rowC'] = {
    params: { param1: 'val3', ... },
    widgets: [Object[], Object[]]
}

// delete an item: stop all the widgets then delete the data
dataModel.byRowId['rowB'].widgets.forEach((w) => { w.stop() })  // or however it is done
delete dataModel.byRowId['rowB']

// set the entry for rowB to null
// allow the rowOrder array to be sparse as the row number is not as important as the order
// of the items in the array
dataModel.rowOrder[dataModel.rowOrder.indexOf('rowB')] = null

// it's also super easy to do
dataModel.rowOrder = dataModel.rowOrder.filter((elem) => { return elem })

Reply:
I thought about that too, and while that makes more logical sense, the filter before sending data up to the configure widget might bog down things and gobble memory. Realistically, eh, I'm guessing we'll have to deal with a couple hundred file path rows at once (more with the CSV version, later), so maybe that's fine.

Reply:
How about we give it a try and revert back to the other method if it proves slow with real world data sets?

Reply:
Sure, I can change that up. I worry it'll slow things down, but it's all in browser, so I bet that'll be negligible. We'll see.

Reply:
Ok, changed over.